LE Audio: Broadcast audio list instead of array

Thalley opened the pull request:
Modifies the broadcast source and sink APIs and implementations to use a slist instead of an array of pointers to hold the streams. 

fixes https://github.com/zephyrproject-rtos/zephyr/issues/43767

Thalley committed:
Bluetooth: Audio: Modify broadcast source to use a list of streams

Modify from array to list. The list approach is easily
implemented as it is used for the unicast groups, and
using the list approach instead of an array does not
require the array to be static.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Modify broadcast sink to use list of streams

Modify the broadcast sink to use a list of streams instead
of an array. This removes the requirement that the array
of stream pointers shall be static, and makes the
broadcast sink API and implementation more similar
to the unicast group and broadcast source.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented:
Depends on https://github.com/zephyrproject-rtos/zephyr/pull/43709

Thalley commented:
Rebased to fix merge conflicts

Thalley commented:
Rebased to fix test failure

MariuszSkamra commented on subsys/bluetooth/audio/broadcast_sink.c:
```diff
@@ -863,7 +858,11 @@ static void broadcast_sink_cleanup_streams(struct bt_audio_broadcast_sink *sink)
 		stream->iso = NULL;
 		stream->group = NULL;
+
+		sys_slist_remove(&sink->streams, NULL, &stream->node);
```
Additionally, we should change the `bt_audio_stream` member `node` node to be prefixed with `_` -> `_node`. So that it would not be used by the user to put it on some other list :D

Thalley replied:
Agreed. I don't think we have any general guidelines for this in Zephyr (none that I am aware of), but the `_` does typically mean "internal". 

We do have a lot (a lot!) of fields in public structs that are internal and do not use the `_` prefix. 

We can easily fix that for ISO and LE Audio still, as those are experimental. Ideally we should be able to do this for all internal fields in public struct, as those should not be used by the applications, but I guess for the stable APIs we should still go through the proper process of changing those (if using `_` as a prefix is a convention we will start to follow

Thalley replied:
Added the "_" prefix for all nodes in LE Audio (not that many yet).

MariuszSkamra commented on subsys/bluetooth/audio/endpoint.h:
```diff
@@ -77,7 +77,7 @@ struct bt_audio_broadcast_source {
 	struct bt_codec_qos *qos;
 	/* The streams used to create the broadcast source */
-	struct bt_audio_stream **streams;
+	sys_slist_t streams;
```
Where's `sys_slist_init`?

Thalley replied:
`sys_slist_init` is defined as 
```c
static inline void sys_slist_init(sys_slist_t *list)
{
	list->head = NULL;
	list->tail = NULL;
}
```

and since the `bt_audio_broadcast_source`s and `bt_audio_broadcast_sink`s are statically defined, as well as `memset` on deletion, `sys_slist_init` provides no value. 

I can, however, still add it as part of the `create` function for semantic reasons if you like.

MariuszSkamra commented on subsys/bluetooth/audio/endpoint.h:
```diff
@@ -96,7 +96,7 @@ struct bt_audio_broadcast_sink {
 	struct bt_iso_chan *bis[BROADCAST_SNK_STREAM_CNT];
 	/* The streams used to create the broadcast sink */
-	struct bt_audio_stream **streams;
+	sys_slist_t streams;
```
ditto

Thalley replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/44828#discussion_r880319450

Thalley committed:
Bluetooth: Audio: User _ prefix for sys_snode_t nodes

The sys_snode_t nodes are used internally, and we use the
convention of the "_" prefix to indicate that this is
an internally used value.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Casper-Bonde-Bose reviewed (dismissed):
LGTM

carlescufi commented:
@Thalley please rebase

Thalley commented:
Rebased to solve merge conflicts

Thalley commented:
Rebased to solve merge conflicts

Thalley commented:
Rebased to solve merge conflicts

carlescufi commented:
@MariuszSkamra please take a look

asbjornsabo reviewed (commented):
I am wondering about the first commit - it does not really do much, and it does not seem to fit neither the title nor the description it has.
Otherwise, this looks sane to me.

(Posting this as a comment, not a "request changes", as I may not be able to follow up and do not want to block the PR.)

Thalley commented:
> I am wondering about the first commit - it does not really do much, and it does not seem to fit neither the title nor the description it has. Otherwise, this looks sane to me.
> 
> (Posting this as a comment, not a "request changes", as I may not be able to follow up and do not want to block the PR.)

Squashed the first commit. Not sure how it ended up like that :D

Thalley commented:
Fixed build error after another commit was merged

Thalley commented:
Rebased to solve merge conflicts

Thalley commented:
Rebased to solve merge conflicts
